### PR TITLE
[DPC-3684] CPI API Gateway requests

### PR DESCRIPTION
**Pull request by chris-ronning-ny**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3684

## 🛠 Changes

This creates the three API requests that are called for in the JIRA ticket. It also sets up the Dockerfile for local development, and imports some CA certs that are required.

## ℹ️ Context for reviewers

(Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.)

## ✅ Acceptance Validation

(How were the changes verified? Did you fully test the acceptance criteria in the ticket? Provide reproducible testing instructions and screenshots if applicable.)

## 🔒 Security Implications

- [ ] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

### CODE REVIEW

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb [hunk 1]
```diff
@@ -4,20 +4,74 @@
...
     client_secret = ENV.fetch('CPI_API_GW_CLIENT_SECRET', nil)
+    cms_idm_url = ENV.fetch('CMS_IDM_OAUTH_URL', nil)
+    @cpi_api_gateway_url = ENV.fetch('CPI_API_GW_BASE_URL', nil)
     @client = OAuth2::Client.new(client_id, client_secret,
-                                 site: 'https://impl.idp.idm.cms.gov/',
-                                 token_url: '/oauth2/aus2151jb0hszrbLU297/v1/token')
+                                 site: cms_idm_url,
+                                 token_url: '/oauth2/aus2151jb0hszrbLU297/v1/token',
+                                 ssl: {
+                                   verify: env != 'local'
```
Comment (chris-ronning-ny) [inline dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:20]: During development, I was finding it impossible to get Ruby to recognize the CA certs that were mounted in Docker. Given that mounting those CA certs did allow CURL to verify the same requests, this appears to be an issue in Ruby. Therefore, in order to unblock development, I've disabled SSL verification in the local environment only. Deployed environments should be using a different URL, and should be fine, but we'll circle back on that once we're ready to continue development.

File: HHS-FPKI-Intermediate-CA-E1.pem [hunk 2]
```diff
@@ -0,0 +1,35 @@
+-----BEGIN CERTIFICATE-----
```
Comment (jdettmannnava) [inline HHS-FPKI-Intermediate-CA-E1.pem:1]: Are the pem files going to be necessary if we don't verify for local (or is that not knowable)?

File: docker-compose.portals.yml [hunk 3]
```diff
@@ -112,6 +112,11 @@ services:
       - DB_USER=postgres
       - DB_PASS=dpc-safe
       - DATABASE_CLEANER_ALLOW_REMOTE_DATABASE_URL=true
+      - CPI_API_GW_CLIENT_ID=insert_client_id_here
```
Comment (jdettmannnava) [inline docker-compose.portals.yml:115]: Should we pass these in as local env variables (like for GOLDEN_MACAROON)?

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb [hunk 4]
```diff
@@ -4,20 +4,74 @@
...
+                                 })
     fetch_token
   end
+  # rubocop:enable Metrics/MethodLength
+
+  def fetch_enrollment(npi)
+    refresh_token
```
Comment (jdettmannnava) [inline dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:27]: I am not sure I like this pattern. I think replacing `@access` with something like 'requester' (I am not so great with names, which would be defined something like:

```
def requester
    fetch_token unless @access.expired?
    @access 
end
```

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb [hunk 5]
```diff
@@ -4,20 +4,74 @@
...
-  attr_accessor :token
+  attr_accessor :access
 
+  # rubocop:disable Metrics/MethodLength
   def initialize
+    env = ENV.fetch('ENV', nil)
     client_id = ENV.fetch('CPI_API_GW_CLIENT_ID', nil)
     client_secret = ENV.fetch('CPI_API_GW_CLIENT_SECRET', nil)
+    cms_idm_url = ENV.fetch('CMS_IDM_OAUTH_URL', nil)
+    @cpi_api_gateway_url = ENV.fetch('CPI_API_GW_BASE_URL', nil)
```
Comment (jdettmannnava) [inline dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:15]: You might want to verify ends-with-slash here

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb [hunk 6]
```diff
@@ -4,20 +4,74 @@
...
+                                 site: cms_idm_url,
+                                 token_url: '/oauth2/aus2151jb0hszrbLU297/v1/token',
+                                 ssl: {
+                                   verify: env != 'local'
+                                 })
     fetch_token
   end
+  # rubocop:enable Metrics/MethodLength
+
+  def fetch_enrollment(npi)
```
Comment (jdettmannnava) [inline dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:26]: Maybe a comment explaining that this returns data about an organization

File: dpc-portal/spec/lib/cpi_api_gw_client_spec.rb [hunk 7]
```diff
@@ -4,16 +4,101 @@
...
+      expect(client_credentials_strategy_instance).to receive(:get_token).and_return(access_token_object_instance)
+      client = CPIAPIGatewayClient.new
+
+      expect(client.access).to eq access_token_object_instance
+    end
+  end
+
+  describe '.fetch_enrollment' do
+    it 'makes a post request' do
+      oauth_client = instance_double(OAuth2::Client)
```
Comment (jdettmannnava) [inline dpc-portal/spec/lib/cpi_api_gw_client_spec.rb:24]: You can move all the shared doubling to a function; would probably make this easier to read.

File: HHS-FPKI-Intermediate-CA-E1.pem (hunk 2 above)
Comment (chris-ronning-ny) [reply HHS-FPKI-Intermediate-CA-E1.pem:1]: I'm 95% positive we'll need the HHS cert in our deployed environment. The Zscaler one, I left there more as a convenience for developers.

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb (hunk 4 above)
Comment (chris-ronning-ny) [reply dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:27]: I like it!

**Comment by chris-ronning-ny** [comment]
@jdettmannnava changes have been applied, please take a look :)

File: dpc-portal/spec/services/cpi_api_gateway_client_spec.rb [hunk 8]
```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require './app/services/cpi_api_gateway_client'
```
Comment (jdettmannnava) [inline dpc-portal/spec/services/cpi_api_gateway_client_spec.rb:4]: Unnecessary with rename

**Review by jdettmannnava — APPROVED** [review]

**Pull request merged**